dna-src/zomes/communities/code/src/communities/mod.rs

```diff
@@ -19,57 +19,92 @@ pub struct Community {
     pub slug: String,
 }
 
+impl Community {
+    pub fn with_address(&self, address: Address) -> CommunityWithAddress {
+        CommunityWithAddress {
+            address,
+            name: self.name.clone(),
+            slug: self.slug.clone()
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug, Clone, DefaultJson)]
+pub struct CommunityWithAddress {
```

Comment:
Is there an inheritance mechanism or something like that for structs so that I could avoid the duplication of

```
pub struct Community {
    pub name: String,
    pub slug: String,
}

pub struct CommunityWithAddress {
    pub address: Address,
    pub name: String,
    pub slug: String
}
```

anything like:

```
CommunityWithAddress extends Community {
    pub address: Address
}
```
?

Reply:
No inheritance, composition only! But I agree this is lame. What about adding this somewhere?
```rust
use serde::ser::Serialize;
use std::fmt::Debug;

#[derive(Serialize, Deserialize, Debug, Clone)]
pub struct EntryWithAddress<T: Serialize + Debug> {
    #[serde(flatten)]
    entry: T,
    address: Address,
}

impl<T: Serialize + Debug> EntryWithAddress<T> {
    pub fn new(entry:T, address: Address) -> Self {
        Self {
            entry,
            address
        }
    }
}

impl<'a, T: Serialize + Debug> From<&'a EntryWithAddress<T>> for JsonString {
    fn from(v: &EntryWithAddress<T>) -> JsonString {
        match ::serde_json::to_string(v) {
            Ok(s) => Ok(JsonString::from_json(&s)),
            Err(e) => {
                eprintln!("Error serializing to JSON: {:?}", e);
                Err(HolochainError::SerializationError(e.to_string()))
            },
        }.expect(&format!("could not Jsonify {}: {:?}", stringify!(EntryWithAddress<T>), v))
    }
}

impl<T: Serialize + Debug> From<EntryWithAddress<T>> for JsonString {
    fn from(v: EntryWithAddress<T>) -> JsonString {
        JsonString::from(&v)
    }
}
```

and then you can write zome functions like:
```rust
pub fn get_community() -> ZomeApiResult<EntryWithAddress<Community>> {
   ...
   Ok(EntryWithAddress::new(community, community_address))
}
```

The `#[serde(flatten)]` does what you expect and makes it look like address is a field of the struct when it is jsonified. Unfortunately DefaultJson doesn't work for structs with type params which is why it has to reimplement the Into<JsonString> traits.

What do you think? I might actually include this in the HDK if it is useful

Reply:
This is great. Glad we can do this. I'm going to merge this branch and make a new PR for that refactor